src/providers/idp/idp_auth_eval.c

```diff
@@ -0,0 +1,160 @@
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <errno.h>
+#include <jansson.h>
+
+#include "util/util.h"
+#include "providers/idp/idp_auth.h"
+#include "krb5_plugin/idp/idp.h"
+#include "providers/krb5/krb5_utils.h"
+
+errno_t eval_device_auth_buf(struct idp_auth_ctx *idp_auth_ctx,
+                             struct pam_data *pd,
+                             uint8_t *buf, ssize_t buflen)
+{
+    uint8_t *user_reply;
+    uint8_t *end;
+    struct sss_idp_oauth2 *user_data = NULL;
+    int ret;
+    struct idp_open_req_data *open_req = NULL;
+
+    user_reply = memchr(buf, '\n', buflen);
+    if (user_reply == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Missing seperator in device auth reply.\n");
+        return EINVAL;
+    }
+
+    *user_reply = '\0';
+    user_reply++;
+
+    end = memchr(user_reply, '\n', buflen - (user_reply - 1 - buf));
+    if (end == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "Missing second seperator in device auth reply.\n");
+        return EINVAL;
+    }
+
+    *end = '\0';
+
+    user_data = sss_idp_oauth2_decode((char *) user_reply);
+    if (user_data == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to decode device auth reply.\n");
+        ret = EINVAL;
+        goto done;
+    }
+
+    ret = attach_oauth2_info_msg(pd, user_data);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "Failed to attach device auth user information.\n");
+        goto done;
+    }
+
+    /* TODO: add first part of buf to a hash so it can be used during
+     * SSS_PAM_AUTHENTICATE. */
+    open_req = talloc_zero(idp_auth_ctx, struct idp_open_req_data);
+    if (open_req == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "Failed to store data to finish authentication.\n");
+        ret = ENOMEM;
+        goto done;
+    }
+
+    open_req->device_code_data = talloc_strdup(open_req, (char *) buf);
+    if (open_req->device_code_data == NULL) {
+        talloc_free(open_req);
+        ret = ENOMEM;
+        goto done;
+    }
+
+    ret = sss_ptr_hash_add(idp_auth_ctx->open_request_table,
+                           user_data->user_code,
+                           open_req, struct idp_open_req_data);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to add device code data [%d][%s].\n",
+                                 ret, sss_strerror(ret));
+        talloc_free(open_req);
+        goto done;
+    }
+
+    ret = EOK;
+done:
+    sss_idp_oauth2_free(user_data);
+
+    return ret;
+}
+
+errno_t eval_access_token_buf(struct idp_auth_ctx *idp_auth_ctx,
+                              struct pam_data *pd, struct sss_domain_info *dom,
+                              uint8_t *buf, ssize_t buflen)
+{
+    int ret;
+    const char *attrs[] = {SYSDB_NAME, SYSDB_UUID, NULL};
+    struct ldb_result *res = NULL;
+    const char *uuid;
+
+    /* TODO: expect access token as well */
```

Comment:
Is this still actual?

Reply:
Yes, this is something I want to add in a later version.